src/sqlfluff/core/templaters/jinja.py

```diff
@@ -116,7 +116,13 @@ def _extract_libraries_from_config(self, config):
             spec = importlib.util.spec_from_file_location(module_name, file_path)
             lib = importlib.util.module_from_spec(spec)
             spec.loader.exec_module(lib)
-            libraries[module_name] = lib
+            if module_name == "__init__":
```

Comment:
So my main thoughts around this PR were around the use of `__init__.py` to store these variables. `__init__.py` has special meaning in Python as in general it indicates the folder to be treated as a package.

So seeing a folder structure like
![image](https://user-images.githubusercontent.com/80432516/144670534-240e692c-605c-47b5-a4d9-00ac81274f52.png)
sort of implies to the user that when they specify in their config
```
[sqlfluff:templater:jinja]
library_path=libs
```
they are importing a package called libs
i.e.
```sql
SELECT 56
FROM {{ libs.foo.schema }}.{{ libs.foo.table("xyz") }}
WHERE {{ libs.bar.equals("x", 23) }} and {{ libs.root_equals("y", 42) }}
```
rather than a series of individual python modules within the libs folder as we are currently doing.

Reply:
Implying package like structure also then implies the inclusion of submodules as well:
e.g.
![image](https://user-images.githubusercontent.com/80432516/144671667-f2f0cb47-259f-48ad-ab4c-7be7fa72ff55.png)
suggests this is possible:
```sql
SELECT 56
FROM {{ libs.submodule.baz.table("xyz") }}
```
which we don't support.

Reply:
We already have root-level functions and variables and I believe it's the best place where we can put such macros and variables

Reply:
Won't nested modules be supported automatically?

Reply:
So I gave some thought to redefining the library to expect a package to expand on the thoughts above.

code would be something like:
```python
def _extract_libraries_from_config(self, config):
        
        library_path = config.get_section(
            (self.templater_selector, self.name, "library_path")
        )
        if not library_path:
            return {}

        libraries = {}
        # We need to determine if the provided library is a folder of
        # python modules or a package (i.e. contains an __init__.py).
        # If it is a package then we can import the library as such,
        # else we import the individual modules.
        package_name = library_path.split("/")[-1]
        init_path = os.path.join(library_path, "__init__.py")
        
        if os.path.exists(init_path):
            # Library is a package.
            spec = importlib.util.spec_from_file_location(package_name, init_path)
            module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(module)
            libraries[package_name] = module

            # Recursively walk submodules and add them to libraries.
            for loader, module_name, _ in pkgutil.walk_packages(
                module.__path__, prefix=f"{package_name}."
            ):
                libraries[module_name] = loader.find_module(module_name).load_module(
                    module_name
                )

        return libraries
```
i.e. import the libs module and then recursively import all of it's modules and submodules.
- Namespace would look something like `{"libs": ..., "libs.foo": ..., "libs.bar": ..., "libs.submodule": ..., "libs.submodule.baz": ...}` 
- However if you try and reference nested modules in Jinja you can't do it. Reason being in `{{ libs.foo.some_func() }}` Jinja tries to search for the attribute `foo` in `libs` whic isn't there, rather than the library path equivalent in Python of doing `import libs.foo; libs.foo.some_func()`. So basically it's not possible and Jinja doesn't support package like libraries (nested or unnested). So we stick with the importing current importing individual modules method (of which `__init__.py` is just considered a standard module which we just apply some unique logic to).

Reply:
> Won't nested modules be supported automatically?

No it doesn't, if you read the lines around your change it:
- lists the content for the files in the supplied directory path
- if the files is a `.py` file load as a standalone module (or if it's a `__init__.py` load the non-private attributes within the module`)
- ignore directories. therefore any subfolders/modules are not included and I don't think you can supply multiple paths to `library_path` in the config.

Reply:
@Tonkonozhenko @barrywhart

So basically in principle I agree with the change, as from my tests above I've verified that a package based approach isn't possible for Jinja.

My main question is:
Are we happy with using `__init__.py` for this custom logic given the potential for user confusion described above. Or could we consider a custom name like `__libroot__.py` to be more clear that this is non-standard logic?

Reply:
@jpy-git WDYT about such an approach for nested modules?
```python
        libraries = {}

        for loader, module_name, is_pkg in pkgutil.walk_packages([library_path]):
            _module = loader.find_module(module_name).load_module(module_name)

            if '.' in module_name:
                *modules, last = module_name.split('.')
                module = reduce(lambda res, path_part: res[path_part], modules, libraries)
                setattr(module, last, _module)
            else:
                libraries[module_name] = _module

        return libraries
```